python/ray/experimental/channel/shared_memory_channel.py

```diff
@@ -480,6 +483,91 @@ def close(self) -> None:
         self._worker.core_worker.experimental_channel_set_error(self._reader_ref)
 
 
+@DeveloperAPI
+class BufferedSharedMemoryChannel(ChannelInterface):
+    """A Channel that allows to buffer inputs.
+
+    If the buffer is full for writes or empty for reads and the operation
+    blocks longer than configured timeouts, RayChannelTimeoutError will be
+    raised.
+
+    As long as the buffer is not full, read/write APIs are non-blocking.
+    The caller of the API should guarantee to consume buffers before they are
+    exhausted.
+    """
+
+    def __init__(
+        self,
+        writer: Optional[ray.actor.ActorHandle],
+        reader_and_node_list: List[Tuple["ray.actor.ActorHandle", str]],
+        num_shm_buffers: int,
+        typ: Optional[Union[int, SharedMemoryType]] = None,
+    ):
+        self._num_shm_buffers = num_shm_buffers
+        self._channels = [
+            Channel(writer, reader_and_node_list, typ) for _ in range(num_shm_buffers)
+        ]
+        # The next index to write from self._channels.
+        self._next_write_index = 0
+        # The next index to read from self._channels.
+        self._next_read_index = 0
+
+    def ensure_registered_as_writer(self):
+        """
+        Check whether the process is a valid writer. This method must be idempotent.
+        """
+        for chan in self._channels:
+            chan.ensure_registered_as_writer()
+
+    def ensure_registered_as_reader(self):
+        """
+        Check whether the process is a valid reader. This method must be idempotent.
+        """
+        for chan in self._channels:
+            chan.ensure_registered_as_reader()
+
+    def write(self, value: Any, timeout: Optional[float] = None) -> None:
+        """Write a value to a channel.
+
+        If the next buffer is available, it returns immediately. If the next
+        buffer is still in use, it blocks until a buffer is available to write.
+        If a buffer is not available within timeout, it raises RayChannelTimeoutError.
+        """
+        # A single channel is not supposed to read and write at the same time.
+        assert self._next_read_index == 0
```

Comment:
QQ: why is the value 0 special here? Should we just check if the write index equals the read index?

Reply:
one channel doesn't call both read/write. So if your channel writes, there's no read

Reply:
I tried this, but I feel like it is not the best way to do it because we don't require to call ensure_* before using read/write technically. So I feel like asserting in read/write is actually better. lmk if you disagree with this comment, I will follow up in other PR